### PR TITLE
feat!: prohibit going out of service_dir in app-service

### PR DESCRIPTION
**Pull request by ValeryAntopol**

### CODE REVIEW

File: crates/fluence-app-service/src/service.rs [hunk 1]
```diff
@@ -144,23 +143,12 @@ impl AppService {
...
-        preopened_files.insert(PathBuf::from(local_dir.clone()));
-        preopened_files.insert(PathBuf::from(tmp_dir.clone()));
+        create(&service_dir.join(SERVICE_LOCAL_DIR_NAME))?;
+        create(&service_dir.join(SERVICE_TMP_DIR_NAME))?;
 
         let mapped_dirs = hashmap! {
-            String::from(SERVICE_LOCAL_DIR_NAME) => PathBuf::from(local_dir),
-            String::from(SERVICE_TMP_DIR_NAME) => PathBuf::from(tmp_dir),
+            SERVICE_LOCAL_DIR_NAME.to_string() => PathBuf::from(SERVICE_LOCAL_DIR_NAME),
+            SERVICE_TMP_DIR_NAME.to_string() => PathBuf::from(SERVICE_TMP_DIR_NAME),
```
Comment (folex) [inline crates/fluence-app-service/src/service.rs:151]: This looks wrong. What is the meaning here? Why are you mapping `"tmp"` to `"tmp"` and `"local"` to `"local"`?

I would expect you to do
```suggestion
            SERVICE_LOCAL_DIR_NAME.to_string() => service_dir.join(SERVICE_LOCAL_DIR_NAME),
            SERVICE_TMP_DIR_NAME.to_string() => service_dir.join(SERVICE_TMP_DIR_NAME),
```

And what about any other directories, eg `/` or `/etc`? If a user writes them to config, they need to be mapped to `service_base_dir` as well.
```
mapped_dirs = { "/will_remove_this_haha" = "/etc" }
```

File: crates/fluence-app-service/src/service.rs [hunk 2]
```diff
@@ -172,7 +160,9 @@ impl AppService {
             module.config.extend_wasi_envs(envs.clone());
             module
                 .config
-                .extend_wasi_files(preopened_files.clone(), mapped_dirs.clone());
+                .extend_wasi_files(<_>::default(), mapped_dirs.clone());
+
+            module.config.set_service_base_dir(&service_dir)
```
Comment (folex) [inline crates/fluence-app-service/src/service.rs:165]: I'm not sure you need custom `mapped_dirs` if you call `set_service_base_dir` anyway 🤔

File: marine/src/config/marine_config.rs [hunk 3]
```diff
@@ -285,12 +304,25 @@ impl<'c> TryFrom<WithContext<'c, TomlWASIConfig>> for MarineWASIConfig {
             Ok((elem.0.into_bytes(), to.into_bytes()))
         };
 
+        let check_path = |path: PathBuf| -> Result<PathBuf, Self::Error> {
+            if path.is_absolute() {
+                return Err(MarineError::InvalidConfig(format!("Absolute paths are not supported in WASI section: {}", path.display())));
```
Comment (folex) [inline marine/src/config/marine_config.rs:309]: I don't understand. Does that mean I cannot do
```
mapped_dirs = { "/tmp" = "/tmp" }
```
anymore?

File: marine/src/config/marine_config.rs [hunk 4]
```diff
@@ -285,12 +304,25 @@ impl<'c> TryFrom<WithContext<'c, TomlWASIConfig>> for MarineWASIConfig {
...
+            }
+
+            if path.components().contains(&Component::ParentDir) {
+                return Err(MarineError::InvalidConfig(format!("Paths containing \"..\" are not supported in WASI section: {}", path.display())));
```
Comment (folex) [inline marine/src/config/marine_config.rs:313]:
```suggestion
                return Err(MarineError::InvalidConfig(format!("Paths containing '{}' are not supported in WASI section: {}", Component::ParentDir, path.display())));
```

File: crates/fluence-app-service/src/service.rs [hunk 5]
```diff
@@ -144,23 +143,12 @@ impl AppService {
...
-        create(&tmp_dir)?;
-
-        let local_dir = local_dir.to_string_lossy().to_string();
-        let tmp_dir = tmp_dir.to_string_lossy().to_string();
-
-        let mut preopened_files = HashSet::new();
-        preopened_files.insert(PathBuf::from(local_dir.clone()));
-        preopened_files.insert(PathBuf::from(tmp_dir.clone()));
+        create(&service_dir.join(SERVICE_LOCAL_DIR_NAME))?;
+        create(&service_dir.join(SERVICE_TMP_DIR_NAME))?;
```
Comment (folex) [inline crates/fluence-app-service/src/service.rs:147]: plz add comment on why you create these directories

File: crates/fluence-app-service/src/service.rs (hunk 1 above)
Comment (ValeryAntopol) [reply crates/fluence-app-service/src/service.rs:151]: Absolute paths and paths containing .. are banned here

https://github.com/fluencelabs/marine/pull/244/files#diff-7ce8317104ac1eb163bd20a0439f963f5eca4c9189c8ca911cc6262748ac6265


And all other mapped directories moved to the service_base_dir

File: crates/fluence-app-service/src/service.rs (hunk 5 above)
Comment (folex) [reply crates/fluence-app-service/src/service.rs:147]: Maybe you don't even need to create them? Maybe they should be created in `set_service_base_dir` function? or after it.

Like, all directories mentioned in the config must be created, I guess.

File: crates/fluence-app-service/src/service.rs (hunk 2 above)
Comment (ValeryAntopol) [reply crates/fluence-app-service/src/service.rs:165]: Yep, changed it to preopened files

File: crates/fluence-app-service/src/service.rs (hunk 5 above)
Comment (ValeryAntopol) [reply crates/fluence-app-service/src/service.rs:147]: Why do so? By creating them i will change the behaviour. I try to add some security wihout changing old behaviour

File: marine/src/config/marine_config.rs (hunk 3 above)
Comment (ValeryAntopol) [reply marine/src/config/marine_config.rs]: Yes. 
You should still be able to do this:
```mapped_dirs = { "/tmp" = "tmp" }```

**Pull request merged**